### PR TITLE
Cypress: Add IMAGE test case

### PR DESCRIPTION
**Pull request by DaiYuzeng**
## Not Implemented Cases

- https://harvester.github.io/tests/manual/images/verify-options-available-for-image/
- https://harvester.github.io/tests/manual/images/negative-upload-invalid-image-file/
- https://harvester.github.io/tests/manual/images/upload-cloud-image/

### CODE REVIEW

File: cypress/pageobjects/image.po.ts [hunk 1]
```diff
@@ -1,45 +1,183 @@
...
+    this.name().input(value?.name);
+    this.url().input(value?.url);
+    this.description().input(value?.description);
   }
 
-  url() {
-    return new LabeledInputPo('.labeled-input', `:contains("URL")`)
+  public delete(name: string) {
+    this.goToList();
+    cy.get('.search').type(name);
```
Comment (noahgildersleeve) [inline cypress/pageobjects/image.po.ts:168]: The selector should be stored as a class member not selected just based on a CSS class
```
private search = '.search';
cy.get(this.search).type(name);


```

File: cypress/pageobjects/image.po.ts [hunk 2]
```diff
@@ -1,45 +1,183 @@
...
+    const image = cy.contains(name);
+
+    image.should('be.visible')
+    image.parentsUntil('tbody', 'tr').click();
+    cy.contains('Delete').click();
+
+    cy.intercept('DELETE', '/v1/harvester/harvesterhci.io.virtualmachineimages/*/*').as('delete');
+    cy.get('.card-container.prompt-remove').contains('Delete').click();
```
Comment (noahgildersleeve) [inline cypress/pageobjects/image.po.ts:176]: The selector should be stored as a class member not selected just based on a CSS class.

File: cypress/pageobjects/image.po.ts [hunk 3]
```diff
@@ -1,45 +1,183 @@
...
   public goToCreate() {
-    this.goToList()
-    cy.contains('Create').click()
+    this.goToList();
+    cy.contains('Create').click();
+  }
+
+  public goToEdit(name: string) {
+    this.goToList();
+    cy.get('.search').type(name);
```
Comment (noahgildersleeve) [inline cypress/pageobjects/image.po.ts:46]: The selector should be stored as a class member not selected just based on a CSS class
```
private search = '.search';
cy.get(this.search).type(name);
```

File: cypress/pageobjects/image.po.ts [hunk 4]
```diff
@@ -1,45 +1,183 @@
...
+    this.addLabels(value);
+    this.save({ upload: true });
   }
 
-  public save() {
-    cy.intercept('POST', '/v1/harvester/kubevirt.io.virtualmachineimage/*').as('createImage');
-    cy.get('.cru-resource-footer').contains('Create').click()
+  public exportImage(vmName: string, imageName: string) {
+    volumes.goToList();
+    cy.get('.search').type(vmName);
```
Comment (noahgildersleeve) [inline cypress/pageobjects/image.po.ts:103]: The selector should be stored as a class member not selected just based on a CSS class
```
private search = '.search';
cy.get(this.search).type(vmName);
```

File: cypress/pageobjects/image.po.ts [hunk 5]
```diff
@@ -1,45 +1,183 @@
...
+
+    volume.should('be.visible')
+
+    volume.parentsUntil('tbody', 'tr').find('td').eq(2).invoke('text').then((volumeName) => {
+      volumes.exportImage(volumeName.trim(), imageName);
+    });
+  }
+
+  public checkState(value: ValueInterface, valid: boolean = true) {
+    cy.get('.search').type(value.name);
```
Comment (noahgildersleeve) [inline cypress/pageobjects/image.po.ts:115]: The selector should be stored as a class member not selected just based on a CSS class
```
private search = '.search';
cy.get(this.search).type(value.name);
```

File: cypress/pageobjects/volume.po.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  public goToCreate() {
+    this.goToList();
+    cy.contains('Create').click();
+  }
+
+  public goToEdit(name: string) {
+    this.goToList();
+    cy.get('.search').type(name);
```
Comment (noahgildersleeve) [inline cypress/pageobjects/volume.po.ts:39]: The selector should be stored as a class member not selected just based on a CSS class

private search = '.search';
cy.get(this.search).type(name);

File: cypress/pageobjects/volume.po.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    const volume = cy.contains(name)
+
+    volume.should('be.visible')
+    volume.click()
+  }
+
+  public exportImage(volumeName: string, imageName: string) {
+    cy.contains(volumeName).parentsUntil('tbody', 'tr').find('.icon-actions').click();
+    cy.get('.list-unstyled.menu').contains('Export Image').click();
+    cy.get('.labeled-input.edit').find('input').type(imageName);
```
Comment (noahgildersleeve) [inline cypress/pageobjects/volume.po.ts:57]: These should have selectors as class members not CSS classes.

File: cypress/pageobjects/volume.po.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    cy.intercept('POST', `v1/harvester/persistentvolumeclaims/*/${ volumeName }?action=export`).as('exportImage');
+    cy.get('.card-actions').contains('Create').click();
```
Comment (noahgildersleeve) [inline cypress/pageobjects/volume.po.ts:60]: These should have selectors as class members not CSS classes.

**Review by noahgildersleeve — CHANGES_REQUESTED** [review]
The tests look great. There's just some updates that we should make to be sure that the code is easy to update with page object model and one linting error. Thanks.

File: cypress/pageobjects/image.po.ts [hunk 9]
```diff
@@ -1,45 +1,183 @@
...
-    cy.get('.cru-resource-footer').contains('Create').click()
+  public exportImage(vmName: string, imageName: string) {
+    volumes.goToList();
+    cy.get('.search').type(vmName);
+
+    const volume = cy.contains(vmName);
+
+    volume.should('be.visible')
+
+    volume.parentsUntil('tbody', 'tr').find('td').eq(2).invoke('text').then((volumeName) => {
```
[the remark this answers is not in the document]
Comment (DaiYuzeng) [reply cypress/pageobjects/image.po.ts:115]: get Volume name by attachedVM
截屏2022-03-25 上午10 46 34

File: cypress/pageobjects/image.po.ts [hunk 10]
```diff
@@ -1,45 +1,183 @@
...
+      volumes.exportImage(volumeName.trim(), imageName);
+    });
+  }
+
+  public checkState(value: ValueInterface, valid: boolean = true) {
+    cy.get('.search').type(value.name);
+    cy.contains(value.name).parentsUntil('tbody', 'tr').find('td.col-image-percentage-bar').contains(valid ? 'Completed' : '0%', { timeout: 300000 }).should('be.visible');
+    cy.contains(value.name).parentsUntil('tbody', 'tr').find('td.col-badge-state-formatter').contains(valid ? 'Active' : 'Failed', { timeout: 300000 }).should('be.visible');
```
[the remark this answers is not in the document]
Comment (DaiYuzeng) [reply cypress/pageobjects/image.po.ts:123]: these are Image state selector and upload/download/export progress bar selector.

截屏2022-03-25 上午10 50 41

**Comment by DaiYuzeng** [comment]
@noahgildersleeve Please take a look at my latest commit, thanks for your review!

File: cypress/pageobjects/image.po.ts (hunk 3 above)
Comment (DaiYuzeng) [reply cypress/pageobjects/image.po.ts]: done

File: cypress/pageobjects/image.po.ts [hunk 11]
```diff
@@ -1,45 +1,183 @@
...
+
+    this.description().input(value?.description);
+    this.addLabels(value);
+    this.save({ edit: true });
+  }
+
+  public createFromUpload(value: ValueInterface) {
+    this.setValue(value);
+    cy.get('.radio-group').contains('File').click();
+    cy.get('#file').selectFile(value.path, { force: true });
```
[the remark this answers is not in the document]
Comment (DaiYuzeng) [reply cypress/pageobjects/image.po.ts]: reseolved

[the remark this answers is not in the document]
Comment (DaiYuzeng) [reply cypress/pageobjects/image.po.ts]: done

File: cypress/pageobjects/image.po.ts (hunk 4 above)
Comment (DaiYuzeng) [reply cypress/pageobjects/image.po.ts]: done

File: cypress/pageobjects/image.po.ts (hunk 5 above)
Comment (DaiYuzeng) [reply cypress/pageobjects/image.po.ts]: done

File: cypress/pageobjects/image.po.ts (hunk 1 above)
Comment (DaiYuzeng) [reply cypress/pageobjects/image.po.ts]: done

File: cypress/pageobjects/image.po.ts (hunk 2 above)
Comment (DaiYuzeng) [reply cypress/pageobjects/image.po.ts]: done

File: cypress/pageobjects/volume.po.ts (hunk 6 above)
Comment (DaiYuzeng) [reply cypress/pageobjects/volume.po.ts]: done

File: cypress/pageobjects/volume.po.ts (hunk 7 above)
Comment (DaiYuzeng) [reply cypress/pageobjects/volume.po.ts]: done

File: cypress/pageobjects/volume.po.ts (hunk 8 above)
Comment (DaiYuzeng) [reply cypress/pageobjects/volume.po.ts]: done

File: cypress/pageobjects/image.po.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (noahgildersleeve) [reply cypress/pageobjects/image.po.ts:115]: Sorry for the confusion, I should have been more clear. I meant a comment in the code where it just says something like
```
// Get volume name from attahched VM in name field of Volume page
volume.parentsUntil('tbody', 'tr').find('td').eq(2).invoke('text').then((volumeName) => {
```

File: cypress/pageobjects/image.po.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (noahgildersleeve) [reply cypress/pageobjects/image.po.ts:123]: Sorry for the confusion, I should have been more clear. I meant a comment in the code where it just says something like

// state indicator for image upload progress percentage bar
cy.contains(value.name).parentsUntil('tbody', 'tr').find('td.col-image-percentage-bar').contains(valid ? 'Completed' : '0%', { timeout: 300000 }).should('be.visible');
// state indicator for status of image upload status e.g. active or uploading
cy.contains(value.name).parentsUntil('tbody', 'tr').find('td.col-badge-state-formatter').contains(valid ? 'Active' : 'Failed', { timeout: 300000 }).should('be.visible');

Also you should probably set the timeouts as a constant from constants.ts

**Review by noahgildersleeve — CHANGES_REQUESTED** [review]

File: cypress/pageobjects/image.po.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (DaiYuzeng) [reply cypress/pageobjects/image.po.ts:116]: done, thanks!

File: cypress/pageobjects/image.po.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (DaiYuzeng) [reply cypress/pageobjects/image.po.ts]: done, thanks!

**Review by noahgildersleeve — APPROVED** [review]
LGTM

**Pull request merged**